README.md
# Magento 2 Dockergento

Plug and play Magento 2 dev environments with docker. **Fastest performance ever** on Mac and Linux.

## Performance Comparison

#### Up to 7x faster development experience on Mac compare to standard docker setups.

<a href="https://www.youtube.com/watch?v=qdUBuDCzHaA&list=PLBt8dizedSZBhcjTL8SM2PS2HEy0mFf5F" target="_blank">
  <img src="docs/img/benchmark_comparison_video.png" alt="Dockergento speed comparison" width="320" height="180" border="5" />
</a>

#### Check out all benchmarks

* [Benchmarks: Dockergento vs Standard Docker](docs/benchmarks.md)

#### Learn more about how that is achieved

* [Overcoming Docker for Mac performance issues](docs/overcome_performance_issues.md)

---

## What is dockergento?

Dockergento is just a bash script ready to use in Linux and Mac to be able to use docker with best native performance.

While performance might no be a problem for Linux, using this tool is the only way you can overcome performance issues on Mac. Dockergento allows you to have different configuration for each system while using the same workflow. So your whole team can work the same way no matter which computer they are using. It just works!

## Supported Systems

* Mac
* Linux

---

## Video Tutorials

If you do not like reading and prefer watching videos. Check out all video tutorials here:

* [Video Tutorials](./docs/video_tutorials.md)

---

## Installation

You only need 3 things on your local machine: `git`, `docker` and `dockergento`

### Install Docker

Follow the installation steps for your system.

<details>
<summary>Mac</summary>
	
1. Install Docker on [Mac](https://docs.docker.com/docker-for-mac/install/)

2. Configure `File Sharing` settings for the folder that contains your projects

	![File Sharing Configuration](docs/img/file_sharing.png)
	
3. Optionally you can also apply these performance tweaks

	* [http://markshust.com/2018/01/30/performance-tuning-docker-mac](http://markshust.com/2018/01/30/performance-tuning-docker-mac)

</details>
	
<details>
<summary>Linux</summary>
	
1. Install docker

	* Install Docker on [Debian](https://docs.docker.com/engine/installation/linux/docker-ce/debian/)
	* Install Docker on [Ubuntu](https://docs.docker.com/engine/installation/linux/docker-ce/ubuntu/)
	* Install Docker on [CentOS](https://docs.docker.com/engine/installation/linux/docker-ce/centos/)

2. Configure permissions
	
	* [Manage Docker as a non-root user](https://docs.docker.com/install/linux/linux-postinstall/)

</details>

### Install dockergento console

1. Clone this repo

    ```
    git clone https://github.com/ModestCoders/magento2-dockergento.git
    ```

2. Add `dockergento` bin into your `$PATH`

    ```
    sudo ln -s $(pwd)/magento2-dockergento/bin/dockergento /usr/local/bin/
    ```
    
3. Open a new terminal tab/window and check that `dockergento` works

	```
	which dockergento
	dockergento
	```

</details>


## Project Setup

Depending the type of project, you can use one of the following setups:

### Dockerize existing project

```
cd <your_project>
dockergento setup
```

### New project

```
mkdir <new_project_name> && cd <new_project_name>
dockergento setup
dockergento create-project
```

### Magento 2 github for contribution
**Disclaimer:** Performance on Mac is slower here due to the huge amount of files in `app` (~20.000 files)

<details>
<summary>Workaround to improve performance on Mac</summary>
	
1. Remove these lines on `docker-compose.dev.mac.yml`
    
    ```
        - ./app:/var/www/html/app:delegated
        - ./dev:/var/www/html/dev:delegated
        - ./generated:/var/www/html/generated:delegated
        - ./pub:/var/www/html/pub:delegated
        - ./var:/var/www/html/var:delegated
    ```
 
2. Sync `app` using `unison` container. Add this in `docker-compose.dev.mac.yml`
     
    ```
    unison:
      volumes:
        - ./app:/sync/app
    ```

3. Mirror not synced folders before executing composer the first time

	```
	dockergento start
	dockergento mirror-host app dev generated pub var
	```

4. If you are editing code in `app`, you need to start unison watcher to sync files between host and container.

	```
	dockergento watch app/code/Magento/<module_name>
	```
    
</details>

```
git clone https://github.com/magento/magento2.git
cd magento2
dockergento setup
```

---

## Usage

### Installing Magento

For a new store with Magento 2.4, you will need to install Magento via the command line before you can use it. Follow the instructions at https://devdocs.magento.com/guides/v2.4/install-gde/install-quick-ref.html, using `db` as the database host and `elasticsearch` as the Elasticsearch host.

For example:

```
 dockergento magento setup:install --base-url=http://local.magento.com --db-host=db --db-name=magento --db-user=magento --db-password=magento \
--admin-firstname=Magento --admin-lastname=User --admin-email=user@example.com \
--admin-user=admin --admin-password=admin123 --language=en_US \
--currency=USD --timezone=America/Chicago --use-rewrites=1 \
--elasticsearch-host=elasticsearch
```

### Start Application

```
dockergento start
dockergento composer install
sudo vim /etc/hosts
// Add -> 127.0.0.1 <your-domain>
```

Open `http://<your-domain>` in the browser 🎉

### Workflow

See detailed documentation about development workflow with dockergento

* [Development Workflow](docs/workflow.md)

---

## More Documentation

* [Multi store configuration](docs/multi_store.md)
* [PHPStorm + Xdebug Setup](docs/xdebug_phpstorm.md)
* [Grumphp setup](docs/grumphp_setup.md)
* [Docker images list](docs/docker_images.md)
* [Other customizations](docs/customizations.md)

## Troubleshooting

* [Troubleshooting](docs/troubleshooting.md)

---

## ChangeLog

* [CHANGELOG.md](CHANGELOG.md)

## Developers

* [Juan Alonso](https://github.com/jalogut)
* [Daniel Lozano](https://github.com/danielozano)
* [Contributors](https://github.com/ModestCoders/magento2-dockergento/graphs/contributors)

## Donations 🙏

We’ve worked very hard to implement this tool. If you find it useful and want to invite us for a beer, just click on the donation button. Thanks! 🍺 

[![Donate](https://img.shields.io/badge/Donate-PayPal-green.svg)](juan.jalogut@gmail.com)

## Resources

This project has been possible thanks to the following resources:

* [docker-magento](https://github.com/markoshust/docker-magento) by [@markshust](https://twitter.com/markshust)
* [Getting Started with Docker for Magento](https://nomadmage.com/product/getting-started-with-docker-for-magento-2/) by [@mostlymagic](https://twitter.com/mostlymagic)
* [Docker Background Sync](https://github.com/cweagans/docker-bg-sync) by [@cweagans](https://twitter.com/cweagans)

## Licence

* [GNU General Public License, version 3 (GPLv3)](http://opensource.org/licenses/gpl-3.0)

## Copyright
(c) ModestCoders


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Version](http://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Added

### Changed

### Fixed

## [4.4.1] - 2021-06-15
### Fixed
 * Check if .gitignore exists when executing dockergento create-project, it may not exist.

## [4.4.0] - 2021-01-04
### Added
  * Added documentation about install command including elasticsearch setup.
  * Added support for XDebug >= 3.

## [4.3.0] - 2020-10-18
### Added
  * Image added php:7.4-fpm for compatibility with Magento versions >= 2.4
### Fixed
  * Increased fastcgi_buffer_size to 32K. This will fix Magento 2.4 max buffer size error.

## [4.2.0] - 2020-05-16
### Added
 * Now can add project name (-COMPOSER_FILE_NAME env var) via project properties.
 * Created .editorconfig file with basic coding style rules.
 * Now you can choose between Magento: Commerce or Open Source when creating a project.

## [4.1.0] - 2020-03-15
### Added
 * Added support for php 7.3

## [4.0.1] - 2019-11-22
### Fixed
 * Linux: permission fix not running on rebuild
 * Linux: now correct permissions are set when UID is not 1000

## [4.0.0] - 2019-09-26
### Added
 * Added elasticsearch service

### Changed
 * Speed up debug-on and debug-off commands
 * Changed naming convention for fpm images in order to have always the last version
 * Updated nginx conf file

### Fixed
 * Fixed large product imports increasing php `max_input_vars` value
 * Fixed link to php image in docs

## [3.5.0] - 08-03-2019
### Added

* Fix permission on linux machines automatically when containers start
* Added bind path `../.composer:/var/www/html/var/composer_home` so composer cache is kept in host.

## [3.4.0] - 13-02-2019
### Added

* Added php version: 7.2.

## [3.3.1] - 01-02-2019
### Change

* Fix bug when executing `debug-on` command on linux machine

## [3.3.0] - 09-11-2018
### Added

* Automatically set `--working-dir=${COMPOSER_DIR}` option for `composer` commands.

## [3.2.0] - 01-11-2018
### Added

* Support for Livereload in google Chrome.

## [3.1.1] - 26-10-2018
### Added

* Documentation updates and Video tutorials section.

## [3.1.0] - 17-10-2018
### Added

* Command descriptions added to `dockergento` console.

## [3.0.0] - 16-10-2018
### Added

* Multiple OS compatibility (Linux, Mac)
* Include `magento2-dockergento-console` directly in this repository.

### Fixed

* Projects with vendor in repository were not working on Mac
* Magento 2 github project was not working on Mac

### Backward incompatible changes

* `dockergento` needs now 2 `docker-compose` files. The base one and the OS specific.
* `app` service has been renamed to `nginx`
* `magento2-dockergento-console` repository has been removed and its code has been added directly into this repo.
* Debug commands have been renamed to `debug-on` and `debug-off`
* `app-volumes` service has been removed.
* New images used for `nginx` and `php`
	* `modestcoders/nginx:1.13`
	* `modestcoders/php:7.1-fpm-1` 

### Upgrade steps
1. Create a backup of your current config:
	
	```
	mv config/dockergento config/dockergento-backup
	mv docker-compose.yml docker-compose.yml.backup
	```

2. Run `dockergento setup` inside your project
3. Edit volumes in `docker-compose.dev.mac.yml` as you had it before. 
4. Edit `nginx` configuration in `config/dockergento/nginx/conf/default.conf`
5. Edit `config/dockergento/properties` if needed.

## [2.1.0] - 21-09-2018
### Changed

* Easier install documentation using `dockergento setup` command
* Set all configuration settings inside `config/dockergento` folder
* Change `docker-compose.yml` filename to `docker-compose.sample.yml` as this file needs to be copied and edited per project.

## [2.0.0] - 13-09-2018
### Changed

* Set whole magento application in a named volume `magento` instead of using 6 volumes:

	```
	app-vendor
	app-generated
	app-var
	pub-static
	pub-media
	integration-test-sandbox
	```

    * Reason: Bug when having volumes inside a file bind mount: [#26157](https://github.com/moby/moby/issues/26157#issuecomment-419722589)

## [1.3.0] - 05-09-2018
### Changed
* Documentation about how to sync `vendor` and `generated` volumes according to latest changes on [magento2-dockergento-console](https://github.com/ModestCoders/magento2-dockergento-console) 

### Removed

* Remove first 2 steps to configure xdebug in PHPStorm are not longer needed because it is enough by just setting the debug port and server configuration
* Remove bind mount of `generated` folder in `unison container. Unison container is now only meant to be used for `watch` command inside vendor modules.

## [1.2.0] - 30-08-2018
### Added

* Development workflow documentation
* Troubleshooting section with info about broken volumes

### Changed

* Use updated node image with right user `app` by default
* Use `modestcoders/php` image
* Xdebug uses now `xdebug.remote_host=host.docker.internal` as it is defined on `modescoders/php` images

### Removed

* Remove not needed `NPM_CONFIG_PREFIX` because default user is now app
* Remove phpstorm xdebug `DBGp` configuration step that is no longer needed

## [1.1] - 24-08-2018
### Changed

* Use `modestcoders/unison:2.51.2` image for unison container
* Update `markoshust/magento-php` images to PHP `7.1` using tag `7.1-fpm-3`
* Update docker setup with fixes for node and integration tests
* Update dependencies and `app-volumes` configuration to avoid issues when `app-volumes` is started several times in same `docker-compose up` execution
* Execute Dockerfile run commands for `app-volumes` in only one layer as it is best practise
* Update README with `node` and `unison` commands to use `run` instead of `exec`
* Update `xdebug + phpstorm` documentation

## [1.0.1] - 15-08-2018
### Changed

* Use `modestcoders/node-php:node8-php7.0` image for node container


console/commands/setup.sh
#!/usr/bin/env bash
set -euo pipefail

DOCKERGENTO_CONFIG_DIR="config/dockergento"

copy_with_consent()
{
    SOURCE_PATH=$1
    TARGET_PATH=$2
    if [[ -e "${TARGET_PATH}" ]]; then
        read -p "overwrite ${TARGET_PATH}? (y/n [n])? " ANSWER_OVERWRITE_TARGET
        if [ "${ANSWER_OVERWRITE_TARGET}" != "y" ]; then
            printf "${RED} Setup interrupted. '${TARGET_PATH}' exists. Please remove it and try again.${COLOR_RESET}\n"
            exit 1
        fi
    fi
    echo " > cp ${SOURCE_PATH} -> ${TARGET_PATH}"
    mkdir -p $(dirname ${TARGET_PATH})
    cp -Rf ${SOURCE_PATH} ${TARGET_PATH}
}

sanitize_path()
{
    PATH_TO_SANITIZE=$1

    SANITIZED_PATH=${PATH_TO_SANITIZE#/}
    SANITIZED_PATH=${SANITIZED_PATH#./}
    SANITIZED_PATH=${SANITIZED_PATH%/}
    echo "${SANITIZED_PATH}"
}

sed_in_file()
{
    SED_REGEX=$1
    TARGET_PATH=$2
    if [[ "${MACHINE}" == "mac" ]]; then
        sed -i '' "${SED_REGEX}" "${TARGET_PATH}"
    else
        sed -i "${SED_REGEX}" "${TARGET_PATH}"
    fi
}

printf "${GREEN}Setting up dockergento config files${COLOR_RESET}\n"
copy_with_consent "${DOCKERGENTO_DIR}/${DOCKERGENTO_CONFIG_DIR}/" "${DOCKERGENTO_CONFIG_DIR}"
copy_with_consent "${DOCKERGENTO_DIR}/docker-compose/docker-compose.sample.yml" "${DOCKER_COMPOSE_FILE}"
copy_with_consent "${DOCKERGENTO_DIR}/docker-compose/docker-compose.dev.linux.sample.yml" "${DOCKER_COMPOSE_FILE_LINUX}"
copy_with_consent "${DOCKERGENTO_DIR}/docker-compose/docker-compose.dev.mac.sample.yml" "${DOCKER_COMPOSE_FILE_MAC}"
#copy_with_consent "${DOCKERGENTO_DIR}/docker-compose/docker-compose.dev.windows.sample.yml" "${DOCKER_COMPOSE_FILE_WINDOWS}"

read -p "Magento root dir: [${MAGENTO_DIR}] " ANSWER_MAGENTO_DIR
MAGENTO_DIR=${ANSWER_MAGENTO_DIR:-${MAGENTO_DIR}}

if [ "${MAGENTO_DIR}" != "." ]; then
	printf "${GREEN}Setting custom magento dir: '${MAGENTO_DIR}'${COLOR_RESET}\n"
    MAGENTO_DIR=$(sanitize_path "${MAGENTO_DIR}")
    printf "${YELLOW}"
    echo "------ ${DOCKER_COMPOSE_FILE} ------"
	sed_in_file "s#/html/var/composer_home#/html/${MAGENTO_DIR}/var/composer_home#gw /dev/stdout" "${DOCKER_COMPOSE_FILE}"
	echo "--------------------"
    echo "------ ${DOCKER_COMPOSE_FILE_MAC} ------"
	sed_in_file "s#/app:#/${MAGENTO_DIR}/app:#gw /dev/stdout" "${DOCKER_COMPOSE_FILE_MAC}"
	sed_in_file "s#/vendor#/${MAGENTO_DIR}/vendor#gw /dev/stdout" "${DOCKER_COMPOSE_FILE_MAC}"
    echo "--------------------"
#    echo "------ ${DOCKER_COMPOSE_FILE_WINDOWS} ------"
#	sed_in_file "s#/app:#/${MAGENTO_DIR}/app:#gw /dev/stdout" "${DOCKER_COMPOSE_FILE_WINDOWS}"
#	sed_in_file "s#/html/app#/html/${MAGENTO_DIR}/app#gw /dev/stdout" "${DOCKER_COMPOSE_FILE_WINDOWS}"
#	sed_in_file "s#/vendor#/${MAGENTO_DIR}/vendor#gw /dev/stdout" "${DOCKER_COMPOSE_FILE_WINDOWS}"
#    echo "--------------------"
    echo "------ ${DOCKERGENTO_CONFIG_DIR}/nginx/conf/default.conf ------"
    sed_in_file "s#/var/www/html#/var/www/html/${MAGENTO_DIR}#gw /dev/stdout" "${DOCKERGENTO_CONFIG_DIR}/nginx/conf/default.conf"
    echo "--------------------"
    printf "${COLOR_RESET}"
fi

read -p "Composer dir: [${COMPOSER_DIR}] " ANSWER_COMPOSER
COMPOSER_DIR=${ANSWER_COMPOSER:-"."}

if [ "${COMPOSER_DIR}" != "." ]; then
    printf "${GREEN}Setting custom composer paths: '${COMPOSER_DIR}'${COLOR_RESET}\n"
    COMPOSER_DIR=$(sanitize_path "${COMPOSER_DIR}")
    printf "${YELLOW}"
    echo "------ ${DOCKER_COMPOSE_FILE_MAC} ------"
	sed_in_file "s#/composer.json#/${COMPOSER_DIR}/composer.json#gw /dev/stdout" "${DOCKER_COMPOSE_FILE_MAC}"
	sed_in_file "s#/composer.lock#/${COMPOSER_DIR}/composer.lock#gw /dev/stdout" "${DOCKER_COMPOSE_FILE_MAC}"
    echo "--------------------"
#    echo "------ ${DOCKER_COMPOSE_FILE_WINDOWS} ------"
#	sed_in_file "s#/composer.json#/${COMPOSER_DIR}/composer.json#gw /dev/stdout" "${DOCKER_COMPOSE_FILE_WINDOWS}"
#	sed_in_file "s#/composer.lock#/${COMPOSER_DIR}/composer.lock#gw /dev/stdout" "${DOCKER_COMPOSE_FILE_WINDOWS}"
#    echo "--------------------"
    printf "${COLOR_RESET}\n"
fi

if [ ! -f "${COMPOSER_DIR}/composer.json" ]; then
    printf "${GREEN}Creating non existing '${COMPOSER_DIR}/composer.json'${COLOR_RESET}\n"
    mkdir -p ${COMPOSER_DIR}
    echo "{}" > ${COMPOSER_DIR}/composer.json
fi
if [ ! -f "${COMPOSER_DIR}/composer.lock" ]; then
    printf "${GREEN}Creating non existing '${COMPOSER_DIR}/composer.lock'${COLOR_RESET}\n"
    echo "{}" > ${COMPOSER_DIR}/composer.lock
fi

read -p "Composer bin dir: [${BIN_DIR}] " ANSWER_BIN_DIR
BIN_DIR=${ANSWER_BIN_DIR:-${BIN_DIR}}

printf "${GREEN}Setting bind configuration for files in git repository${COLOR_RESET}\n"
add_git_bind_paths_in_file()
{
    GIT_FILES=$1
    FILE_TO_EDIT=$2
    SUFFIX_BIND_PATH=$3

    BIND_PATHS=""
    while read FILENAME_IN_GIT; do
        if [[ "${MAGENTO_DIR}" == "${FILENAME_IN_GIT}" ]] || \
            [[ "${MAGENTO_DIR}" == "${FILENAME_IN_GIT}/"* ]] || \
            [[ "${FILENAME_IN_GIT}" == "vendor" ]] || \
            [[ "${FILENAME_IN_GIT}" == "${DOCKER_COMPOSE_FILE%.*}"* ]]; then
            continue
        fi
        NEW_PATH="./${FILENAME_IN_GIT}:/var/www/html/${FILENAME_IN_GIT}"
        BIND_PATH_EXISTS=$(grep -q -e "${NEW_PATH}" ${FILE_TO_EDIT} && echo true || echo false)
        if [ "${BIND_PATH_EXISTS}" == true ]; then
            continue
        fi
        if [ "${BIND_PATHS}" != "" ]; then
            BIND_PATHS="${BIND_PATHS}\\
      " # IMPORTANT: This must be a new line with 6 indentation spaces.
        fi
        BIND_PATHS="${BIND_PATHS}- ${NEW_PATH}${SUFFIX_BIND_PATH}"

    done <<< "${GIT_FILES}"

    printf "${YELLOW}"
    echo "------ ${FILE_TO_EDIT} ------"
    sed_in_file "s|# {FILES_IN_GIT}|${BIND_PATHS}|w /dev/stdout" "${FILE_TO_EDIT}"
    echo "--------------------"
    printf "${COLOR_RESET}"
}

if [[ -f ".git/HEAD" ]]; then
    GIT_FILES=$(git ls-files | awk -F / '{print $1}' | uniq)
    if [[ "${GIT_FILES}" != "" ]]; then
        add_git_bind_paths_in_file "${GIT_FILES}" "${DOCKER_COMPOSE_FILE_MAC}" ":delegated"
#        add_git_bind_paths_in_file "${GIT_FILES}" "${DOCKER_COMPOSE_FILE_WINDOWS}" ""
    else
        echo " > Skipped. There are no files added in this repository"
    fi
else
    echo " > Skipped. This is not a git repository"
fi

echo "PHP version:"
DEFAULT_PHP_VERSION="7.1"
AVAILABLE_PHP_VERSIONS="7.0 7.1 7.2 7.3 7.4"
select PHP_VERSION in ${AVAILABLE_PHP_VERSIONS}; do
    if $(${TASKS_DIR}/in_list.sh "${PHP_VERSION}" "${AVAILABLE_PHP_VERSIONS}"); then
        break
    fi
    if $(${TASKS_DIR}/in_list.sh "${REPLY}" "${AVAILABLE_PHP_VERSIONS}"); then
        PHP_VERSION=${REPLY}
        break
    fi
    echo "invalid option '${REPLY}'"
done

printf "${GREEN}Setting php version: '${PHP_VERSION}'${COLOR_RESET}\n"
if [ "${PHP_VERSION}" != "${DEFAULT_PHP_VERSION}" ]; then
    printf "${YELLOW}"
    echo "------ ${DOCKER_COMPOSE_FILE} ------"
	sed_in_file "s#php:${DEFAULT_PHP_VERSION}#php:${PHP_VERSION}#gw /dev/stdout" "${DOCKER_COMPOSE_FILE}"
	sed_in_file "s#php${DEFAULT_PHP_VERSION}#php${PHP_VERSION}#gw /dev/stdout" "${DOCKER_COMPOSE_FILE}"
    echo "--------------------"
    printf "${COLOR_RESET}\n"
fi

read -p "[OPTIONAL] Docker compose project name: [${COMPOSER_PROJECT_NAME-}]" ANSWER_PROJECT_NAME
PROJECT_NAME=${ANSWER_PROJECT_NAME-}
if [[ ! -z "${PROJECT_NAME}" ]];then
    printf "${GREEN}Setting custom project name: '${PROJECT_NAME}'\n"
    printf "This will serve as docker object (container, volume, network) prefix.\n"
    printf "${COLOR_RESET}"
    COMPOSE_PROJECT_NAME="${PROJECT_NAME}"
fi

printf "${GREEN}Saving custom properties file: '${DOCKERGENTO_CONFIG_DIR}/properties'${COLOR_RESET}\n"
cat << EOF > ./${DOCKERGENTO_CONFIG_DIR}/properties
MAGENTO_DIR="${MAGENTO_DIR}"
COMPOSER_DIR="${COMPOSER_DIR}"
BIN_DIR="${BIN_DIR}"
COMPOSE_PROJECT_NAME="${COMPOSE_PROJECT_NAME}"
EOF

# Stop running containers in case that setup was executed in an already running project
${COMMANDS_DIR}/stop.sh

echo ""
printf "${YELLOW}-------- IMPORTANT INFO: -----------${COLOR_RESET}\n"
echo ""
echo "   Docker bind paths were automatically added here:"
echo ""
echo "      * ${DOCKER_COMPOSE_FILE_MAC}"
#echo "     * ${DOCKER_COMPOSE_FILE_WINDOWS}"
echo ""
echo "   Please check that they are right or edit them accordingly."
echo "   Be aware that vendor cannot be bound for performance reasons."
echo ""
printf "${YELLOW}-------------------------------------${COLOR_RESET}\n"

echo ""
printf "${GREEN}Dockergento set up successfully!${COLOR_RESET}\n"
echo ""
